### ISSUE TITLE
[ENHANCEMENT] Multi-modal tensors combining

### ISSUE THREAD

**Issue by wanliAlex** [marqo-ai/marqo#317]
**Is your feature request related to a problem? Please describe.**
Currently, we don't have a way to combine the embedded tensors from multiple fields into one.
We would like an endpoint to allow users to combine multiple fields into one tensor.

**Describe the solution you'd like**
A detailed design document will be provided.

**Comment by wanliAlex** [comment]
# Multi-modal Tensors Combining Design Document
## Introduction

Currently, for a document with the following format in tensor search at indexing,
```python
document = {"image_field_1" : "https://image-fieled-1.png", #v1
		   "image_field_2" : "https://image-fieled-2.png", #v2
		   "Text_field_1" : "text-field-1-random-random", #v3
		   "Text_field_2" : "text-field-2-random-random", #4
		   } 
```
we will need to inference each field once, and generate **4** tensors for this document. We denote the 4 tensors as $v_1, v_2, v_3, v_4$, respectively.

In search, 
```python
query = "i want to search something" #q1
```
we will vectorise the query into tensor as $q_1$ and compare the distance with the above 4 tensors to get the `_score`.
```math
score = \max \{ distance(v_i, q_1) | i \in \{ 1,2,3,4 \} \},
```
where $distance(\cdot,\cdot)$ is a *__linear__* distance measurement method (e.g., dot product).

In this design, we would like to find a way to combine several multi-modal fields into one. For example, for the
document mentioned above, we add a new field called `multi-modal-combination':
```python
document = {"image_field_1" : "https://image-fieled-1.png", #v1
		   "image_field_2" : "https://image-fieled-2.png", #v2
		   "text_field_1" : "text-field-1-random-random", #v3
		   "text_field_2" : "text-field-2-random-random", #v4
		   "multimodal_combination" : {
                                         "image_field_1": w1, #a float number,
					 "text_field_1": w3, #a float number,
					 "text_field_2": w4, #a float number} # v^*
		   } 
```

For this document, we will need to do inference 4 times to compute $v_1, v_2, v_3, v_4$ and compute the multi-modal-combination tensor $v^*$ as:
```math
v^* = (w_1 *v_1 + w_3*v_3 + w_4 *v_4) / (w_1 +w_3 +w_4)
```
and for the same search query, the score will then be computed as:
```math
score = \max \{distance(v_2, q_1), distance(v^*, q_1) \}.
```
This is the **Option 2** combination strategy in the later section, which **depreciates the searchability of those combined fields**
 ($v_1, v_3, v_4$ in this case). 

The other **option 1** is to **reserve the searchability of those fields**, where the score is computed as 
```math
score = \max \{d_1, d_2, d_3, d_4, d^* \}.
```

## Design 

The multi-modal tensor combination happens at the indexing time (`mq.index("test").add_documents()`). We can add a new parameter `multimodal_combination: Dict[str, float] = None ` to achieve the feature. The `str` should be the fields that are expected to be combined in the documents, and the `float` number is the corresponding weight.

### Client

#### Basic Constraints
- At the client, the first constraint is that there shouldn't be overlapping between `non_tensor_fields` and `multimodal_combination` .
- The second constraint is the parameter can only be used with the index settings `"normalize_embeddings":  True` .
- There should be more than 1 fields (>=2) as there is no point to combine one field.

#### Combination Strategy
One additional parameter can be `multimodal_combination_replace: bool = True` which defines the strategy of the multimodal-combination. The `multimodal_combination` field is used to represent the combined tensors. Now there are two options to deal with the combined fields:

- 1. Reserve the searchability of those fields. The combined fields will still be used in a search query. This will increase the storage and time complexity of Marqo. However, it will provide better search performances, especially if there is an exact match with the query. With this strategy, the document will be treated as:
	```python
	document = {"image_field_1" : v1,
			   "image_field_2" : v2,
			   "text_field_1" : v3,
			   "text_field_2" : v4,
			   "multimodal_combination" : v^*,
			   } 
	```
- 2. Depreciate the searchability of those fields. The combined fields will not be used in tensor search but represented by the `multimodal-combination` field. This reduces the storage and time complexity but degrades search performances. With this strategy, the document will be treated as:
	```python
	document = {#"image_field_1" : v1,
			   "image_field_2" : v2,
			   #"text_field_1" : v3,
			   #"text_field_2" : v4,
			   "multimodal_combination" : v^*,
			   } 
	```

#### Example
```python
mq.index("my-first-index").add_documents([
		{  "Title":  "The Travels of Marco Polo",  
		"Description":  "A 13th-century travelogue describing the travels of Polo",  
		"Genre":  "History"  },  
		{"Title":  "Extravehicular Mobility Unit (EMU)",  
		"Description":  "The EMU is a spacesuit that provides environmental protection",  
		"_id":  "article_591",  
		"Genre":  "Science"  }],  
		multimodal_combination= {"Title": 0.3, "Description" : 0.7},
		multimodal_combination_replace = True,
		)
```

### Marqo
The feature should be added after the inference of all the fields ([link](https://github.com/marqo-ai/marqo/blob/1d9d563c4d87a7fbed03d6e212d5d917768c761f/src/marqo/tensor_search/tensor_search.py#L404-L573)).

#### Combination Strategy
Based on the **combination strategy** , we should have different strategies to deal with those combined fields.

#### Combination Fields and Documents Fields
There might be a mismatch between combination fields and the actually fields in the documents. For example, the combination field has 3 fields but the documents only have 2 of them. There are two strategies here:

- 1. Raise an error for this strategy for this document. This should not stop the add_documents process but just add a failure state for this specific documents.
- 2. We do the multimodal-combination for those two fields. That is to say we always do a **union** between the combination fields and the actual document fields. We combine the fields as long as the **union** returns a set that is larger than **2**


#### Confusion with search weight boost
Note that this feature should not be confused with the search weight boost feature. The boost affects the score computation process (**query time**), while the multimodal-combination affects the vectorise process (**indexing time**). We can even boost the `multimodal_combiation_field` if necessary.

**Comment by pandu-k** [comment]
I prefer combination strategy (i) because preventing tensorising of the source fields can be implemented by  `non_tensor_fields`. This keeps things simpler, especially because the behaviour of `multimodal_combination_replace: bool = True` might be confusing during an add_and_update call (users might remove existing tensors from documents).

**Comment by jn2clark** [comment]
@pandu-k i tend to agree with you. the only thing is, what are the implications of forcing people to store the original fields?

**Comment by pandu-k** [comment]
@jn2clark I agree we should limit the number of tensor fields, but I think this problem can be solved a simpler way. Something like an inverse `non_tensor_field` argument: `tensor_fields` (return error if request has both parameters as not null)

for this doc we could do: `tensor_fields=["multi_modal_combination", "image_field_2"]`

**Comment by pandu-k** [comment]
Also, do we want to have the flexibility of more than 1 combination field?

**Comment by wanliAlex** [comment]
@pandu-k just to clarify, for `tensor_fields=["multi_modal_combination", "image_field_2"]`, fields `"image_field_1"`, `"text_field_1"`, `"text_field_2"` will still be vectorised as they are in `"multi_modal_combination"`.

**Comment by wanliAlex** [comment]
I think it's possible to have multiple combination fields.

**Comment by pandu-k** [comment]
@wanliAlex but they won't be stored (saving storage) 

> @pandu-k just to clarify, for `tensor_fields=["multi_modal_combination", "image_field_2"]`, fields `"image_field_1"`, `"text_field_1"`, `"tenxt_field_2"` will still be vectorised as they are in `"multi_modal_combination"`.

**Comment by pandu-k** [comment]
> I think it's possible to have multiple combination fields.

What would the API look like?

**Comment by wanliAlex** [comment]
```python
mq.index("my-first-index").add_documents([
		{  "Title":  "The Travels of Marco Polo",  
		"Description":  "A 13th-century travelogue describing the travels of Polo",  
		"Genre":  "History"  },  
		{"Title":  "Extravehicular Mobility Unit (EMU)",  
		"Description":  "The EMU is a spacesuit that provides environmental protection",  
		"_id":  "article_591",  
		"Genre":  "Science"  }],  
		multimodal_combination= [{"name": "combined_1", "Title": 0.3, "Description" : 0.7},
                                         {"name": "custom_name_2", "Title": 0.5, "Description" : 0.5}]
		multimodal_combination_replace = True,
		)
```

@pandu-k would it be good to provide a `"name"` field in the dictionary?

**Comment by jn2clark** [comment]
> Also, do we want to have the flexibility of more than 1 combination field?

absolutely

**Comment by pandu-k** [comment]
What about this type of structure: 
```
.add_documents([
		{  "Title":  "The Travels of Marco Polo",  
		"Description":  "A 13th-century travelogue describing the travels of Polo",  
		"Genre":  "History"  },  
		{"Title":  "Extravehicular Mobility Unit (EMU)",  
		"Description":  "The EMU is a spacesuit that provides environmental protection",  
		"_id":  "article_591",  
		"Genre":  "Science"  }, 
                {"combo": {
                       "https://image_store.my_image.png": 0.4,
		       "What is a dog":  -2},
		 "Title: "Not dogs"
                }
], non_tensor_fields=["Genre", "Description"])
```

**Comment by pandu-k** [comment]
We don't want to do any chunking

**Comment by wanliAlex** [comment]
I am finishing a POC so we can discuss it during the meeting.

**Comment by wanliAlex** [comment]
If we add the documents like this
```python
        tensor_search.add_documents(config=self.config, index_name=self.index_name_1, docs=[
            {
                "Title": "Horse rider",
                "text_field": "A rider is riding a horse jumping over the barrier.",
                "image_field": "https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg",
                "_id": "0"

            },
            # {
            #     "Title": "red bus",
            #     "text_field": "A red bus is running on the street with a lot of passengers inside.",
            #     "image_field": "https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image4.jpg",
            #     "_id": "1"
            # }
        ], auto_refresh=True, multimodal_combination = [{"image_field": 0.5, "text_field":0.5}])
```

### The chunks without this feature:
```python
[{'__vector_Title': 'This is the very long vector for the field. [0,....0]',
  '__field_content': 'Horse rider',
  '__field_name': 'Title',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_text_field': 'This is the very long vector for the field. '
                         '[0,....0]',
  '__field_content': 'A rider is riding a horse jumping over the barrier.',
  '__field_name': 'text_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_image_field': 'This is the very long vector for the field. '
                          '[0,....0]',
  '__field_content': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg',
  '__field_name': 'image_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'}]
```

### The chunks with this feature
```python
[{'__vector_Title': 'This is the very long vector for the field. [0,....0]',
  '__field_content': 'Horse rider',
  '__field_name': 'Title',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_text_field': 'This is the very long vector for the field. '
                         '[0,....0]',
  '__field_content': 'A rider is riding a horse jumping over the barrier.',
  '__field_name': 'text_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_image_field': 'This is the very long vector for the field. '
                          '[0,....0]',
  '__field_content': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg',
  '__field_name': 'image_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_multimodal_field_0': 'This is the very long vector for the field. '
                                 '[0,....0]',
  '__field_content': "{'image_field', 'text_field'}",
  '__field_name': "{'image_field', 'text_field'}",
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'}]
```

@jn2clark @pandu-k any comments?

**Comment by pandu-k** [comment]
@Jeadie Suggests this format (based on my suggestions):
```
add_documents([
    {
         "title": "Cats and Dogs",
         "my combo field": {
             "https://favouriteimagestore/dog_image.png": {
               "weight": 0.1,
               "newParam": "ModelA",
             }
             "Caring for a pet requires a lot of effort": {...
         }
    }
])
```

**Comment by wanliAlex** [comment]
```python
add_documents([
    {
         "title": "Cats and Dogs",
         "my combo field": {
             "https://favouriteimagestore/dog_image.png": {
               "weight": 0.1,
               "newParam": "ModelA",
             },
             "Caring for a pet requires a lot of effort": {
              "weight":0.9,
         }
      }
    }
])
```

**Comment by wanliAlex** [comment]
@pandu-k 

With the documents:
```python
        tensor_search.add_documents(config=self.config, index_name=self.index_name_1, docs=[
            {
                "Title": "Horse rider",
                "combo_text_image": {
                    "A rider is riding a horse jumping over the barrier." : {
                        "weight" : 0.5,
                                        },
                    "https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg": {
                    "weight": 0.5,
                                        },
                },
                "_id": "0"

            },
        ], auto_refresh=True)
```

This is the generated chunks:
```python
[{'__vector_Title': 'This is the very long vector for the field. [0,....0]',
  '__field_content': 'Horse rider',
  '__field_name': 'Title',
  'Title': 'Horse rider',
  'combo_text_image': ['A rider is riding a horse jumping over the barrier.',
                       'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg']},
 {'__vector_combo_text_image': 'This is the very long vector for the field. '
                               '[0,....0]',
  '__field_content': ['A rider is riding a horse jumping over the barrier.',
                      'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'],
  '__field_name': 'combo_text_image',
  'Title': 'Horse rider',
  'combo_text_image': ['A rider is riding a horse jumping over the barrier.',
                       'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg']}]
```

For another test:
add document:
```python
            {
                "Title": "Horse rider",
                "text_field": "A rider is riding a horse jumping over the barrier.",
                "image_field":"https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg",
                "_id": "1"
            },
```
chunks:
```python
[{'__vector_Title': 'This is the very long vector for the field. [0,....0]',
  '__field_content': 'Horse rider',
  '__field_name': 'Title',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_text_field': 'This is the very long vector for the field. '
                         '[0,....0]',
  '__field_content': 'A rider is riding a horse jumping over the barrier.',
  '__field_name': 'text_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'},
 {'__vector_image_field': 'This is the very long vector for the field. '
                          '[0,....0]',
  '__field_content': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg',
  '__field_name': 'image_field',
  'Title': 'Horse rider',
  'text_field': 'A rider is riding a horse jumping over the barrier.',
  'image_field': 'https://raw.githubusercontent.com/marqo-ai/marqo/mainline/examples/ImageSearchGuide/data/image1.jpg'}]
```

**Comment by jn2clark** [comment]
looks good @wanliAlex . @pandu-k what filtering or lexical search operations are available on the meta-data when it is stored as a list?

**Comment by pandu-k** [comment]
> looks good @wanliAlex . @pandu-k what filtering or lexical search operations are available on the meta-data when it is stored as a list?

lexical search, filtering (for both lexical and tensor search)

**Comment by jn2clark** [comment]
thanks @pandu-k . I am pretty happy with this

**Comment by jn2clark** [comment]
merged in #332
